apps/ejabberd/src/ejabberd_admin.erl

```diff
@@ -325,6 +337,86 @@ registered_users(Host) ->
     SUsers = lists:sort(Users),
     lists:map(fun({U, _S}) -> U end, SUsers).
 
+-spec import_users(Filename :: string()) -> [{ok, ejabberd:user()} |
+                                             {exists, ejabberd:user()} |
+                                             {not_allowed, ejabberd:user()} |
+                                             {invalid_jid, ejabberd:user()} |
+                                             {null_password, ejabberd:user()} |
+                                             {bad_csv, binary()}].
+import_users(File) ->
+    FileStream = stdio:file(File),
+    CsvStream = csv:stream(FileStream),
+    Workers = spawn_link_workers(),
+    WorkersQueue = queue:from_list(Workers),
+    do_import(CsvStream, WorkersQueue).
+
+-spec do_import(CsvStream :: stdio:stream(), Workers :: queue:queue()) ->
+    [{ok, ejabberd:user()} |
+     {exists, ejabberd:user()} |
+     {not_allowed, ejabberd:user()} |
+     {invalid_jid, ejabberd:user()} |
+     {null_password, ejabberd:user()} |
+     {bad_csv, binary()}].
+do_import({}, WorkersQueue) ->
+    Workers = queue:to_list(WorkersQueue),
+    lists:flatmap(fun get_result/1, Workers);
+
+
+do_import({s, UserData, TailFun}, WorkersQueue) ->
+    {{value, Worker}, Q1} = queue:out(WorkersQueue),
+    Worker ! {proccess, UserData},
+    Q2 = queue:in(Worker, Q1),
+    do_import(TailFun(), Q2).
+
+-spec spawn_link_workers() -> [pid()].
+spawn_link_workers() ->
+    [ spawn_link(?MODULE, register_worker, [self()]) ||
+      _ <- lists:seq(1, ?REGISTER_WORKERS_NUM)].
+
+-spec get_result(Worker :: pid()) -> [{ok, ejabberd:user()} |
+                                      {exists, ejabberd:user()} |
+                                      {not_allowed, ejabberd:user()} |
+                                      {invalid_jid, ejabberd:user()} |
+                                      {null_password, ejabberd:user()} |
+                                      {bad_csv, binary()}].
+get_result(Pid) ->
+    Pid ! get_result,
+    receive
+        {result, Result} -> Result
+    end.
+
+-spec register_worker(Manager :: pid()) -> ok.
+register_worker(Manager) ->
```

Comment:
The function name is misleading. It suggest that it will register a worker in some registry, while in fact this is the main function of a worker registering users.